subprojects/core/src/main/java/org/gradle/internal/classpath/types/ExternalPluginsInstrumentationTypeRegistry.java

```diff
@@ -44,38 +46,67 @@ public ExternalPluginsInstrumentationTypeRegistry(Map<String, Set<String>> direc
 
     @Override
     public Set<String> getSuperTypes(String type) {
-        // TODO: This can probably be optimized: optimize it if it has performance impact
         Set<String> superTypes = Collections.emptySet();
         if (type.startsWith(GRADLE_CORE_PACKAGE_PREFIX)) {
             superTypes = gradleCoreInstrumentationTypeRegistry.getSuperTypes(type);
         }
-        return !superTypes.isEmpty() ? superTypes : Collections.unmodifiableSet(computeAndCacheSuperTypesForExternalType(type));
+
+        if (!superTypes.isEmpty()) {
+            return superTypes;
+        } else {
+            superTypes = instrumentedSuperTypes.get(type);
+            return superTypes != null
+                ? superTypes
+                : computeAndCacheInstrumentedSuperTypes(type, new HashSet<>());
+        }
     }
 
-    private Set<String> computeAndCacheSuperTypesForExternalType(String type) {
+    private Set<String> computeAndCacheInstrumentedSuperTypes(String type, Set<String> visited) {
         // We intentionally avoid using computeIfAbsent, since ConcurrentHashMap doesn't allow map modifications in the computeIfAbsent map function.
         // The result is, that multiple threads could recalculate the same key multiple times, but that is not a problem in our case.
         Set<String> computedInstrumentedSuperTypes = instrumentedSuperTypes.get(type);
         if (computedInstrumentedSuperTypes == null) {
-            computedInstrumentedSuperTypes = computeInstrumentedSuperTypes(type);
+            computedInstrumentedSuperTypes = computeInstrumentedSuperTypes(type, visited);
             instrumentedSuperTypes.put(type, computedInstrumentedSuperTypes);
         }
         return computedInstrumentedSuperTypes;
     }
 
-    private Set<String> computeInstrumentedSuperTypes(String type) {
+    private Set<String> computeInstrumentedSuperTypes(String type, Set<String> visited) {
+        // In case of detected type cycle, calculate super types without recursive cache
+        Set<String> superTypes = visited.add(type)
+            ? computeSuperTypesWithRecursiveCaching(type, visited)
+            : computeSuperTypesWithoutRecursiveCaching(type);
```

Comment:
💭 This the main change: when we detect a cycle (a type that was already visited), we change the strategy and discover it's subtypes without recursion and without caching subtrees.